ENH: Short-Time Fourier Transform (STFT) Enhancements

vollbier opened the pull request:
I found working with the [stft](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html) / [istft](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.istft.html#scipy.signal.istft) functions to be a bit tedious, so I would like to propose a more versatile class [ShortTimeFFT](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L56), which provides a parametrized short-time Fourier transform and its associated inverse. E.g.: 
```python    
SFT = ShortTimeFFT(win=np.hanning(50), hop=10, T=1e-3, scale_to='magnitude')
Sx = SFT.stft(x)
y = SFT.itsft(Sx, k1=n_x)  # parameter k1 determines last signal index of y
```   
Using a class made it easy to provide additional features, like the members [lower_border_end](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L1122) and [upper_border_begin](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L1136), which can be used to determine the part of the STFT, which is affected by border effects, as shown in the following plot: It shows the STFT of a 50 s sine signal with varying frequency $f_i(t)$ with the shaded areas on the left and the right indicating border effects caused by the window slices in that area not fully being inside time range of the signal (marked by vertical dashed lines):

![STFT_example](https://user-images.githubusercontent.com/12721170/201481721-88508219-ae4c-41c3-9b9c-95164fd5e76d.png)
The code for the plot is in the `ShortTimeFFT` documentation. 


# Algorithmic Enhancements

The algorithm implemented in the [istft](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.istft.html#scipy.signal.istft) function, i.e., the overlap-add reconstruction, conceals a little gem: The underlying least squares solution can be rearranged in a way so that each STFT point `(t,f)` can be interpreted as a kind of base function (named [dual_win](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L794)) being shifted by time t and modulated by frequency `f`. E.g.,
```python
T, N = 0.1, 51
d_win = gaussian(N, std=1/T, sym=True)  # symmetric Gaussian window
SFT = ShortTimeFFT.from_dual(d_win, h_, T)
Sx = SFT.stft(x)
```
decomposes the signal `x` into modulated and shifted Gaussians by utilizing [from_dual](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L249). Having an explicit time-frequency series expansion of the signal makes reasoning about it much more intuitive. E.g., the dual window makes it clear how to generalize the concept of constant-overlap perfect reconstruction filters for [FFT filter banks](https://en.wikipedia.org/wiki/Filter_bank#FFT_filter_banks). Since I am only aware of discussions of the dual windows in the more general context of the quite abstract [Frame theory](https://en.wikipedia.org/wiki/Frame_(linear_algebra)), I included a simpler [derivation](https://github.com/vollbier/scipy/blob/shortTimeFFT/doc/source/tutorial/signal.rst#inverse-stft-and-dual-windows) only requiring some understanding of manipulating matrices in the Scipy User Guide.

The ability to work with signal and STFT parts is another new feature: Assuming `x` is `N` samples long, the SFT of the first and second signal half can be determined with
```python
    Sx = SFT.stft(x)  # full signal
    p_q = SFT.nearest_k_p(N // 2)
    Sx0 = SFT.stft(x[:p_q])  # lower half
    Sx1 = SFT.stft(x[p_q:])  # upper half
```
When assembling sequential signal parts together, the overlap needs to be considered:
```python
    p0_ub = SFT.upper_border_begin(p_q)[1] - SFT.p_min
    p1_le = SFT.lower_border_end[1] - SFT.p_min
    Sx01 = np.hstack((Sx0[:, :p0_ub],
                      Sx0[:, p0_ub:] + Sx1[:, :p1_le],
                      Sx1[:, p1_le:]))
    np.allclose(Sx01, Sx)  # Compare with complete STFT => True
```
It is also possible to calculate the `itsft` for signal parts:
```python
    y_p = SFT.istft(Sx, N//3, N//2)
    np.allclose(y_p, x[N//3:N//2])  # => True
```
Note that this ability to consistently work with signal and STFT slices highlights a difference in the old and new implementation: To reconstruct the original signal the new [istft](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L835) implementation expects the time slice range to be [p_min](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L1033):[p_max](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L1084), i.e., being larger the time range of the signal (as shown in the plot above). The old version assumes the first slice always to be centered at t=0. The new approach has also further benefits:

* When formulating the STFT as a series expansion, the original stft() does not guarantee to use the same dual window on the borders as on the inner signal, which makes interpreting it in those regions cumbersome.
* The new version fulfills the [NOLA condition](file:///home/dietrich/GitHub/scipy/doc/build/html/reference/generated/scipy.signal.check_NOLA.html) more robustly, since there are more slices available at the edges. This is relevant when using windows which are not nonzero everywhere.
* The behavior is explicitly documented (see the Short-Time Fourier Transform section in the Scipy User Guide).


## Other new features

* Provide the scaling factors [fac_magnitude](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L922) and [fac_psd](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L943) to convert a calculated STFT to magnitude or psd scaling. By introducing the [scale_to](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L471) method, the scaling behavior is also easier to understand.
* Provide the FFT types 'onesided' and 'onesided2X' to make it transparent when and how the STFT values are doubled when utilizing a one-sided spectrum. This [comment](https://github.com/scipy/scipy/issues/14903#issuecomment-1100249704) in #14903  discussed the challenges of the old parametrization.
* Provide a `phase_shift` parameter to avoid be able to avoid phase skew (new default behavior). Closes #9026.
* [spectrogram](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L703) calculates the absolute square of the STFT as well a cross-spectrogram (closes #10914). Having this together with [stft](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L597) avoids concealing the subtle differences in parameterization. Closes #14903 and allows PR #16000 to be closed).

## Unit tests (~~TODO~~ done)

The main task left is to complete the unit tests. Currently, the `ShortTimeFFT` is [shimmed](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/tests/_scipy_spectral_test_shim.py#L1) into the current [stft/istft tests](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/tests/test_spectral.py#L15). It demonstrates that the `ShortTimeFFT` is a super set of  features of the current implementation and that the results are comparable. This solution is quite hacky and should probably be removed once `ShortTimeFFT` has sufficient unit testing.

Furthermore, the [csd](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_spectral_py.py#L464) function contains a [shimmed](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_spectral_py.py#L636) `_spectral_helper()` function (which implements the actual STFT) for testing. The original aim was to provide an implementation to be able to remove the [_spectral_helper](https://github.com/scipy/scipy/blob/v1.9.3/scipy/signal/_spectral_py.py#L1616-L1907) function. Since the [csd](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.csd.html) is not improved, and it is not guaranteed that it behaves identically with all possible parametrizations, it is probably not that beneficial. Hence, it is planned to remove those shims, when the `ShortTimeFFT` unit tests are complete. Currently, they are still included to provide additional testing.

## Documentation (~~TODO~~ done)
While documenting, I could not figure the following things out:

1. It would be helpful to be able to enlarge the matplotlib plots a bit (e.g., [here](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/scipy/signal/_short_time_fft.py#L168) by around 30% or so). Then the size of the label font would fit better to the size of the plot window.
2. I used [Inkscape](https://inkscape.org) to create this [drawing](https://raw.githubusercontent.com/vollbier/scipy/d9a7b918af31179dc819af86df377059c8ac9d38/doc/source/_static/tutorial_stft_sliding_win_startstop.svg). The current firefox does not display the arrow tips in the SVG. Hence, I included a [png file](https://github.com/vollbier/scipy/blob/d9a7b918af31179dc819af86df377059c8ac9d38/doc/source/_static/tutorial_stft_sliding_win_start.png) instead. Is there a better way? Furthermore, I am not sure, if the best place for graphics in the User manual is [_static](https://github.com/vollbier/scipy/tree/d9a7b918af31179dc819af86df377059c8ac9d38/doc/source/_static) directory. Also, I did not figure out how to include a separate version for the dark theme.
3. Is there a way to have all the class method documentations on the same page and not on separate pages?
4. Doing `python dev.py refguide-check` on the doctest comment
    ```doctest
    >>> ax1.set(title="A nice Plot")
    ```
    generates an error, whereas
    ```doctest
    >>> ax1.set_title("A nice Plot")
    ```
    does not. I am not sure if it is a matplotlib or a `refguide-check` problem.

Any help here would very be much appreciated.


## Additional affected Issues

* The suggestion in #14740 of providing ``norm=None``, i.e., not scaling the window function is implemented.
* #13577 ask for to works with chunks for Dask - this is possible with the current implementation as shown above.
* #15963 can be closed due to this [comment](https://github.com/scipy/scipy/issues/15963#issuecomment-1098427351).


## Summary
Due to redoing and enhancing the underlying STFT algorithm, a new API to was implemented to expose the resulting benefits. Mainly:

* Providing a new explicit interpretation of the iSTFT as decomposition of the input signal into an expansion of a time and frequency shifted function named "dual window".
* Allow to consistently slice and reassemble consecutive signal and STFT chunks.
* Provide the time regions of the STFT which are influenced by the border effects.
* A more robust iSTFT implementation.
* Expanded and improved documentation.

The main task to complete are the unit tests. Furthermore, the shimmed tests show that the original stft() / istft() as well as the [spectrogram](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.spectrogram.html) function can be deprecated without the loss of functionality. It makes sense to keep the [_spectral_helper](https://github.com/scipy/scipy/blob/v1.9.3/scipy/signal/_spectral_py.py#L1616-L1907) function  just for the [csd](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.csd.html) implementation.

I would love feedback, if the `ShortTimeFFT` implementation could be a useful addition for SciPy. Especially comments about the clarity of the docstrings would be helpful. Of course any further support on improving this PR is very welcome.

## Update: To-dos are finished
  All open to-do items of this post have been completed. The main differences to the initial committed are noted [here](https://github.com/scipy/scipy/pull/17408#issuecomment-1374256593).

vollbier commented:
Thanks for taking the time to give feedback, @larsoner. Pinging past `stft()` contributors is a good idea—friendly ping to  @e-q, @gauss256, @WarrenWeckesser @rgommers, @ewmoore, @NickleDave, @endolith.

The suggestion of porting the [MNE Python test](https://github.com/mne-tools/mne-python/blob/main/mne/time_frequency/tests/test_stft.py) was quite helpful. Besides revealing a bug on multidimensional inputs, it shows why [_spectral_helper()](https://github.com/scipy/scipy/blob/v1.9.3/scipy/signal/_spectral_py.py#L1805) preserves the signal energy poorly (as in #14903): The first time slice is always at `t=0`, which can lead to "missing" components, which do not necessarily carry signal information but signal energy (due to the `istft()` implementation). The `ShortTimeFFT` includes by default all slices which "touch" the signal, hence preserving the energy (as tested [here](https://github.com/vollbier/scipy/blob/shortTimeFFT/scipy/signal/tests/test_short_time_fft.py#L234)).

This also calls to investigate the [csd](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.csd.html) function - it probably suffers from the same problem. The easiest thing to do might be to port it into `ShortTimeFFT` and deprecate the original. Since `csd()` only gets used by `welch()` - that function would have to be adapted.

Concerning code duplication: Since `_spectral_helper()` is only used by `spectrogram()`, `stft()` and `csd()`, I shimmed  the `ShortTimeFFT` for unit testing. In the light of the comment above, IMHO, the sensible thing to do is to port `csd()` to `ShortTimeFFT`, deprecate `spectrogram()`, `stft()`, `csd()` and then remove `_spectral_helper()`.

Commit [ceceaaa](https://github.com/scipy/scipy/pull/17408/commits/ceceaaaf3d6c4d3dcfde4c12a481667f344052a8) fixes the problems with the figure size in docstrings and the arrow tips problems in the SVGs.

larsoner commented:
> the sensible thing to do is to port csd() to ShortTimeFFT, deprecate spectrogram(), stft(), csd() and then remove _spectral_helper().

Could we keep spectrogram, csd, and stft as ~1-10 line wrappers around calls to the new class? If it's that easy then it's worth keeping them around and noting in the docstring that they are legacy. The maintenance cost to us is near zero if it's that easy, and people can transition code to the new interface whenever they want.

vollbier commented:
For testing I already implemented [ _stft_wrapper](https://github.com/vollbier/scipy/blob/bc3f491928cffc2d0cde9b1056409d3ecc8d558a/scipy/signal/tests/_scipy_spectral_test_shim.py#L37), [_istft_wrapper](https://github.com/vollbier/scipy/blob/bc3f491928cffc2d0cde9b1056409d3ecc8d558a/scipy/signal/tests/_scipy_spectral_test_shim.py#L150) and [ _spect_helper_csd](https://github.com/vollbier/scipy/blob/bc3f491928cffc2d0cde9b1056409d3ecc8d558a/scipy/signal/_spectral_py.py#L665) . Unfortunately it was quite fidgety, to make these wrappers behave like the original, making them a bit longer than 10 lines. I worry a bit that they do not behave 100% identically and thus cause confusion. Especially explaining the parts of the existing implementation which are not completely documented, like the padding behavior (see energy conservation), might cause quite a bit of additional maintenance work.   

Wouldn't it be easier to

- mark the existing implementation as deprecated over a reasonable amount of time (4 years or so) before removing it,
- provide a porting example in the documentation and
- perhaps add a convenience method for wrapping `get_window()`?

This would push the burden of dealing with the pathological edge cases onto the user.

larsoner commented:
> I worry a bit that they do not behave 100% identically and thus cause confusion. Especially explaining the parts of the existing implementation which are not completely documented, like the padding behavior (see energy conservation), might cause quite a bit of additional maintenance work.

So you are saying it would be a lot of work / lines of code to get your new class to reproduce the `_spectral_helper` outputs (to numerical precision)? Or that it's maybe not a lot of work, but what we have right now could be considered incorrect?

vollbier commented:
Numerical accuracy is a always challenging topics. E.g., this [test](https://github.com/scipy/scipy/pull/17408/checks?check_run_id=9602571000) using`istft_wrapper()`  fails on Arm64  on `assert_allclose(atol=2e-15)` but runs with `atol=1e-15` (floating point resolution) on my machine. I think that it is quite possible that such deviations will occur in other places as well, making someone's unit test fail. ...
Concerning the existing unit tests, the test wrappers produce the same output (on my machine)—I am not really convinced that those tests cover all pathological cases.

On correctness:

Regarding the problems with energy conservation: I consider the missing slices (and thus missing energy) to be a bug (#15704 is incorrect in that regard—I was not aware of that problem at the time).  Those missing slices also allow to design pathological windows (which have zero values at the right places), where the old `istft()` fails but the new one does not.

Furthermore, I think that not all combinations of the `stft()` parameters ` scaling, mode, return_onesided` (as shown [here](https://github.com/scipy/scipy/issues/14903#issuecomment-1100249704)) make sense. I.e., there is not straightforward way to interpret the scaling. If for an application only the relative values matter, this is of course not a problem—but porting it to the `ShortTimeFFT` would need a workaround, as those insensible combination are not implemented.

endolith commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1421 @@
+    the signal. If only the dual window is known and the STFT is invertible,
+    `from_dual` can be used to instanciate this class.
+
+    Due to the convetion of time t = 0 being at the first sample of the input
```
```suggestion
    Due to the convention of time t = 0 being at the first sample of the input
```

endolith commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1421 @@
+    `from_dual` can be used to instanciate this class.
+
+    Due to the convetion of time t = 0 being at the first sample of the input
+    signal, the SFT values typically has negative time slots. Hence,
```
```suggestion
    signal, the STFT values typically have negative time slots. Hence,
```

Unless "SFT" stands for something else?

endolith commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1421 @@
+    More detailed information can be found in the :ref:`tutorial_stft` section
+    of the :ref:`user_guide`.
+
+    Note that all parameters of the initializer, expect `scale_to` (which uses
```
```suggestion
    Note that all parameters of the initializer, except `scale_to` (which uses
```

Is this what was meant?

endolith commented:
This looks really great!

Link to docs here FYI: https://output.circle-artifacts.com/output/job/7c3cce65-22dd-47bc-ad5f-ca9880b7f4a2/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.html

* Could it use `fs` instead of `T`?  `fs` is usually an exact whole number, unlike `T`, and is used by the `scipy.signal.stft` function, and at least in the audio field is the more obvious number.
* Typos I put in the PR

I'll look more later

vollbier commented:
Thanks for pointing out the typos, @endolith. 

Concerning if the sampling frequency `fs` or the sampling interval `T=1/fs` is in this context the more natural quantity to use:
- I gravitate towards `T`. The reason being , `T`  and `delta_t` correspond to the x-axis of the STFT and it is not confused as easily with `delta_f` (the y-axis). 
- I added a property `fs` (getter and setter), to show the connection to `T`.
- `T` is only used to calculate times slice and frequency bin values, i.e., in the methods `delta_t`, `delta_f`, `t()`, `f` as well as the power spectral density scaling factor `fac_psd`—so it does not have any palpable influence on the accuracy of the STFT.
- I guess, writing `ShortTimeFFT(w, hop=10, T=1/48e3)` or `ShortTimeFFT(w, hop=10, fs=48e3)`  is a matter of habit. Perhaps there is somebody else would like to express their opinion on this matter. I'm fine with following suit to the taste of the majority on  the `__init()__` parameters (the only place beside the property where `T` is user visible).

larsoner commented:
> Could it use fs instead of T? fs is usually an exact whole number, unlike T, and is used by the scipy.signal.stft function, and at least in the audio field is the more obvious number.

I'm inclined toward `fs` as well. For me personally it's more natural / I see it more often, and I think for `scipy.signal` more generally we've tried to stick with it as what we're standardizing around.

> Regarding the problems with energy conservation: I consider the missing slices (and thus missing energy) to be a bug

Okay then in principle I think we can fix this bug with no need for a deprecation period

> Furthermore, I think that not all combinations of the stft() parameters scaling, mode, return_onesided (as shown https://github.com/scipy/scipy/issues/14903#issuecomment-1100249704) make sense. I.e., there is no straightforward way to interpret the scaling. If for an application only the relative values matter, this is of course not a problem—but porting it to the ShortTimeFFT would need a workaround, as those insensible combination are not implemented.

If it's not too much work to support these (I mean like < 100 lines or something) then I think I'd prefer to do it. But if the modes somehow give a numerically incorrect result we shouldn't allow them.

@e-q added `spectorgram` 8 years ago and stft [7 years ago](#6058) so it's likely that there is a lot of existing code out there that uses these functions.

vollbier commented:
Making some more progress:

* Added a new classmethod `from_window`, which wraps `get_window`, to be more aligned with the traditional   functions. One can do now
  ```Python
  SFT = ShortTimeFFT.from_window('hann', fs=100, nperseg=128, noverlap=64, 'onesided')
  Sx = SFT.stft(x)
  ```  
* Changed `T` to `fs` in `__init__()` and classmethods.
* Improved documentation a bit, especially providing a link to the `ShortTimeFFT` on its child HTML pages.
* Wrappers of `stft`, `istft` and `csd` for unit testing are now all in file  [_scipy_spectral_test_shim.py](https://github.com/vollbier/scipy/blob/shortTimeFFT/scipy/signal/tests/_scipy_spectral_test_shim.py). 

Regarding replacing `_spectral_helper()` - I am still unsure: The wrappers are 350 lines of code - `_spectral_helper()` has roughly 200 (without docstr). So it's a bit above the 100 lines @larsoner suggested. If `_spectral_helper()` stays, the wrappers can be removed in the  future, i.e., when we are confident that `test_short_fft.py` test thoroughly enough.

vollbier commented:
Thanks @endolith, @larsoner and @ilayn for the input so far! I finally completed my open tasks, so I think this PR is now ready for review. The main differences to the initial commit are:

- [x] Completed unit tests and documentation.
- [x] Implemented [wrappers](https://github.com/scipy/scipy/blob/e01ae4d6580436db53c271bcb3372ddc1a74221e/scipy/signal/tests/_scipy_spectral_test_shim.py) of `stft`, `istft` and `csd` and added shims to unit tests to verify results of old and new implementations are close. These wrappers show that replacing `_spectral_helper` will cause compatability problems in some edge cases.
- [x] Provided a `spectrogram` method.
- [x] Ported time interval parameter `T` to sampling frequency `fs` in initializers and class-methods.
- [x] Provided class-method `ShortTimeFFT.from_window(win_param, fs, nperseg, noverlap, ...)` wrapping `get_window` for better compatability with existing implementation.
- [x] Added [comparison](https://output.circle-artifacts.com/output/job/e3da5d8b-b252-43d2-8ae1-8a1ec8c09329/artifacts/0/html/tutorial/signal.html#comparison-wit-classic-implementation) subsection of old and new implementation to user guide.

Rendered documentation: [ShortTimeFFT](https://output.circle-artifacts.com/output/job/e3da5d8b-b252-43d2-8ae1-8a1ec8c09329/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.html#scipy.signal.ShortTimeFFT) and [Short-Time Fourier Transform section in the user guide](https://output.circle-artifacts.com/output/job/e3da5d8b-b252-43d2-8ae1-8a1ec8c09329/artifacts/0/html/tutorial/signal.html#short-time-fourier-transform
). Note that the links pointing to rendered code in this post are only valid while commit [e01ae4d](https://github.com/scipy/scipy/pull/17408/commits/e01ae4d6580436db53c271bcb3372ddc1a74221e) is the last one.

vollbier commented:
Gentle ping to @endolith, @larsoner and @ilayn: Anything more I can do advance this PR - I think it is ready for review (as stated in the previous [comment](https://github.com/scipy/scipy/pull/17408#issuecomment-1374256593))?

PS: The failing tests from the stats module are unrelated to this PR.

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1666 @@
+
+Implementation Notes (as of 2022-12)
+------------------------------------
+* When the minimal version of Python is bumped to 3.9 then
+  ``@lru_cache(maxsize=None)`` can be replaced with ``@cache`` from functools,
+  which should be marginally faster.
+* MyPy Version 0.931 does not support decorated property methods. Hence,
+  applying ``@property`` to methods decorated with `@lru_cache(maxsize=None)``
+  causes a MyPy error. Hence, the caching in method `lower_border_end()` is
+  done by hand. MyPy Version 0.950 does not seem to have this limitation.
+* Since `NDArray` is only available for Numpy versions 1.21 and above, the
+  methods `stft()`, `stft_detrend()`, `istft()`, and  `_ifft_func()` need to
+  use `NDArray` array as a parameter instead of `NDArray[complex]`.
```
Just a fly by comment reading these 😃 We bumped all the dependencies here, minimal Python is now 3.9, Mypy is 1.0 and NumPy is 1.21.6.

vollbier replied:
Thanks @tupui. I updated my  dependencies—I am still not quite best friends with mypy yet 😐

larsoner commented:
@vollbier sorry for the (very) long delay here. I'll add it to my TODO list for this week

larsoner commented on scipy/signal/tests/_scipy_spectral_test_shim.py:
```diff
@@ -0,0 +1,483 @@
+    if x.dtype in (np.float32, np.complex64):
+        Sxx = Sxx.astype(np.complex64)
+
+    # workaround for test_average_all_segments() - seems to be buggy behavior:
```
😬

vollbier replied:
Are referring to the casting or the the comment? :shrug:

larsoner commented on scipy/signal/tests/test_spectral.py:
```diff
@@ -9,10 +9,15 @@
+                          spectrogram, check_COLA, check_NOLA)
 from scipy.signal._spectral_py import _spectral_helper
 
+# Compare ShortTimeFFT.stft() / ShortTimeFFT.stft() with stft() / istft():
+from ._scipy_spectral_test_shim import stft_compare as stft
+from ._scipy_spectral_test_shim import istft_compare as istft
+from ._scipy_spectral_test_shim import csd_compare as csd
```
Instead of doing this, can we add a new `test_spectral_stft.py` with contents:
```
# Duplicate test_spectral tests, but using newer (I)STFT functions
from test_spectral import *
from ._scipy_spectral_test_shim import stft_compare as stft
from ._scipy_spectral_test_shim import istft_compare as istft
from ._scipy_spectral_test_shim import csd_compare as csd
```
? I *think* that would make it so that both types of tests run now and in the future.

tupui replied:
Also let's use absolute imports please.

vollbier replied:
I already ported all relevant tests from `test_spectral.py` to `test_short_time_fft.py` by hand. The original purpose of the test shims was to have an easy way of cross-checking during implementing. In my opinion, the test shims do not add any benefit to unit testing (anymore). I left them in, since they nicely document the more subtle differences of the implementation choices. I would be fine with removing them.

larsoner reviewed (commented):
Sorry for the long delay, finally looked and it seems very well tested and documented!

I think we should use the `.. legacy::` directive for stft/istft:

https://scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#legacy-directive

Other than that it's just a bunch of minor stuff. I think we should merge this sooner rather than later so that people can start using it!

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+from typing import Callable, get_args, Literal, Optional, Union
+
+import numpy as np
+from numpy.typing import NDArray
```
For consistency with the rest of our typing, I would instead use `npt.ArrayLike` or `np.ndarray` depending on the context.

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+
+
+# noinspection PyShadowingNames
+class ShortTimeFFT:
```
There is a long list of methods/attributes, could you add a list in the docstrings? Also do we really need all of that? It's easy to add something in the public API, but very hard to remove something once it's released. It's ok to "plan" a bit ahead (and keep a few things a bit private even), but we need to be conservative and mindful about the extent of the public API.

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+"""Short-time Fourier Transform Module.
+
+Implementation Notes (as of 2023-04)
+------------------------------------
```
These should go into an issue or as comments instead. I think here these will end up in the public documentation which is not wanted.

vollbier replied:
Since `_short_time_fft.py` is not a module, it does not appear in the public documentation.

tupui replied:
Right, but that's still not a good place for it and we had cases like this were such comments just stay there and are not being used/updated. Our workflow revolves mostly around issues so I think this should go there.

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+  ``NDArray[np.complex128]``). Hence the methods `stft()`, `stft_detrend()`,
+  `istft()`, and  `_ifft_func()` need to use `NDArray` array as a parameter
+  instead of `NDArray[complex]`.
+* The entry ``.. currentmodule:: scipy.signal.ShortTimeFFT`` is required
+  in docstrings to ensure that `stft` and `istft` generate HTML links to
+  methods and not the legacy functions. The side effect is that `ShortTimeFFT`
+  does not work anymore. It is not known how to do this file-wide instead of
+  method-wide.
```
I am not sure I see the issue here, but I don't think we need to do this in other places.

vollbier replied:
Writing `` `stft` `` in a docstring generates a link to the legacy implementation. Using ``.. currentmodule:: scipy.signal.ShortTimeFFT`` circumvents this.

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+  methods and not the legacy functions. The side effect is that `ShortTimeFFT`
+  does not work anymore. It is not known how to do this file-wide instead of
+  method-wide.
+* The HTML documentation currently renders each method/property on a separate
+  page without reference to the parent class. Thus, a link to `ShortTimeFFT`
+  was added to the "See Also" section of each method/property.
```
This should not be needed, see other examples of classes for how we document these usually. e.g. `scipy.stats.qmc.Sobol`.

vollbier replied:
On the page [scipy.stats.qmc.Sobol.random](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.qmc.Sobol.random.html#scipy.stats.qmc.Sobol.random) I do not see a link to the parent class [scipy.stats.qmc.Sobol](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.qmc.Sobol.html). Such a link would simplify navigating around a bit, IMO.

tupui replied:
It might, but that's a larger issue and should not be dealt with here. It's more something to be done on the theme side itself.

That's a general thing, some choices and suggestions are made might not be "optimal", but we need to find a balance between "fixing" things and being consistent with the rest of the code base. Otherwise we end up with multiple ways of dealing with a problem and we are struggling then to maintain things.

tupui commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+  methods and not the legacy functions. The side effect is that `ShortTimeFFT`
+  does not work anymore. It is not known how to do this file-wide instead of
+  method-wide.
+* The HTML documentation currently renders each method/property on a separate
+  page without reference to the parent class. Thus, a link to `ShortTimeFFT`
+  was added to the "See Also" section of each method/property.
```
This should not be needed, see other examples of classes for how we document these usually. e.g. `scipy.stats.qmc.Sobol`.

tupui commented on scipy/signal/tests/test_short_time_fft.py:
```diff
@@ -0,0 +1,775 @@
+from scipy.signal.windows import gaussian
+
+
+def chk_VE(match):
+    """Assert for a ValueError matching regexp `match`.
+
+    This little wrapper allows a more concise code layout. Usage::
+
+        with chk_VE("Message RegExp"):
+            test_code
+            ...
+    """
+    return pytest.raises(ValueError, match=match)
```
I would not do this. For maintenance this complicate our task as the code does not look familiar compared to other places.

tupui reviewed (changes requested):
Thank you @vollbier for the extensive work 🚀 

I am not an expert in the matter, but I still have a few comment on the API. AFAICT, this is a replacement of `stft` and `istft`, hence if we don't want this class to also follow the same path, we need to be sure about the public API, the flexibility for future enhancements and the maintainability.

vollbier commented on scipy/signal/_spectral_py.py:
```diff
@@ -699,6 +700,11 @@ def spectrogram(x, fs=1.0, window=('tukey', .25), nperseg=None, noverlap=None,
+    The class `ShortTimeFFT` is a newer STFT / ISTFT implementation with more
+    features also including a spectrorgam method. A :ref:`comparison
+    <tutorial_stft_classic_stft>` between the implementations can be found in
+    the :ref:`tutorial_stft` section of the :ref:`user_guide`.
```
Thanks, fixed in commit 3b0c4a69e9060decf59ca794e97bd36f83964c36.

vollbier commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+                 mfft: Optional[int] = None,
+                 dual_win: Optional[NDArray] = None,
+                 scale_to: Optional[Literal['magnitude', 'psd']] = None,
+                 phase_shift: Optional[int] = 0):
```
Good Point. Renamed `fft_type` to `fft_mode` (left the `fft_` part in, since `mode` is too generic and already used in the legacy implementation). In commit 126e8309a7f40ede82d5ea208f7dac178e659e51, I added the `*` parameter to `__init__`, `from_dual`, `from_window`, `stft`, `stft_detrend`, `spectrogram`, and `istft`.

vollbier commented on scipy/signal/__init__.py:
```diff
@@ -283,8 +283,11 @@
+   ShortTimeFFT   -- Interface for calculating the \
+                     :ref:`Short Time Fourier Transform <tutorial_stft>` and \
+                     its inverse.
+   stft           -- Compute the Short Time Fourier Transform (legacy).
+   istft          -- Compute the Inverse Short Time Fourier Transform (legacy).
```
In its header, I did declare the file `_short_time_fft.py` as a "module", though it only provides the class `ShortTimeFFT`. This is fixed in commit 126e8309a7f40ede82d5ea208f7dac178e659e51 (The confusion stems of a more liberal interpretation of the term [module](https://docs.python.org/3.11/glossary.html#term-module) by me as is customary of SciPy).

vollbier commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+# Linter does allow to import ``Generator`` from ``typing`` module:
+from collections.abc import Generator
+from functools import cache, lru_cache, partial
+from typing import Callable, get_args, Literal, Optional, Union
```
That would be nice, but the union operator was [introduced](https://docs.python.org/3.10/whatsnew/3.10.html#pep-604-new-type-union-operator) in Python 3.10 but the [development release notes](https://scipy.github.io/devdocs/release/1.11.0-notes.html) require only Python 3.9.
Is that not correct?

vollbier commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+        return self._hop
+
+    @property
+    def T(self) -> float:
```
In signal processing, the sampling interval `T` appears frequently as a variable (e.g., [here](https://docs.scipy.org/doc/scipy/tutorial/fft.html)). When having different `T`s, it can make the code more legible, when its used context is clear. Furthermore, since `ShortTimeFFT` is not a `ndarray`, but always a scalar, mistaking it for a matrix transpose is not really a problem, IMO.

vollbier commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+        """
+        return self._win
+
+    @property
+    def hop(self) -> int:
```
They make the user experience smoother:

- The attributes `win`, `dual_win` and `hop` are immutable, since they cannot be modified independently of each other. 
- IMO, there are applications, where it makes sense to mutate the other attributes (e.g., a GUI for an interactive spectrogram).
- The setters do parameter checking and provide friendly error messages. I feel that performing the parameter checking in separate setter methods instead of doing it all in the initializer makes the code less cluttered. 
- The mostly primitive getter methods are the obvious downside of this approach, which are not really user-visible. Hence, they are primarily a mild annoyance for the reviewers :grin:

vollbier commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+
+
+# noinspection PyShadowingNames
+class ShortTimeFFT:
```
For classes, the list of methods and attributes are automatically generated by Sphinx (like [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.qmc.Sobol.html)). In the method and attribute docstrings, I used the "See Also" section to link to related methods/attributes. 
What would be the benefit of an additional list in the docstrings?

vollbier replied:
I know at first glance, the API seems a bit excessive, but I pondered quite a bit over it:

- As a general note: In the wonderful world of signal processing the algorithms are usually comparatively simple, but the [reasoning](https://doi.org/10.1007/978-1-4612-0003-1) about them is not. Hence, the underlying idea of this API is to provide some aide in thinking about the STFT in a consistent manner. 
- The new section in the user guide explains the reasons for many of those methods (especially the indexing).
- This API has three immutable and five mutable attributes, which are all natural to the STFT. Getting the indexing right (as laid out in the user's guide) is also a bit of chore. Hence, the methods quickly add up. I looked through the class again and did not find anything which appeared superfluous to me. 
- I do get the point that removing something from a public API is very difficult (which is the motivation for this rewrite). But the majority of the utility methods do not modify the instance. Hence, the risk of locking down future features is not that high.

vollbier commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1659 @@
+    r"""Provide a parametrized discrete Short-time Fourier transform (stft)
+    and its inverse (istft).
+
+    .. currentmodule:: scipy.signal.ShortTimeFFT
```
I found by trial and error that I needed this for the links within the class to work. Did something change in the Sphinx configuration?

h-vetinari commented:
> > @larsoner: I think we should merge this sooner rather than later so that people can start using it!
> 
> @tupui: @larsoner I would not rush anything. The branch cut is at the end of this month. At minima we need to nail the public API if we want to really push. I am a bit concern here with the size of the PR. Did you have time to review everything? I only see a small discussion.

I agree that this is getting rather tight for 1.11. Is there a good reason we couldn't take a bit more time and get it into 1.12 early? API changes later on are very painful, so I'd like us to be quite-to-very sure that those won't be necessary.

larsoner commented:
> I am a bit concern here with the size of the PR. Did you have time to review everything? I only see a small discussion.

I did look through the code a bit but I'm mostly convinced by the extensive unit tests, including some I've had sitting around for ~10 years that never worked with `stft` but do work with this PR. @endolith also looked back in Nov 22 and was happy so that gives me some additional confidence.

> I agree that this is getting rather tight for 1.11. Is there a good reason we couldn't take a bit more time and get it into 1.12 early?

Just that we have to pull the trigger at some point because I think this PR brings a lot of useful stuff to the table. But if @rgommers would like to take a look or people think there should be additional iterations/clarifications (like @tupui you had quite a few comments and should probably check to see if they've all been addressed since commits have been pushed) then it's okay to move the milestone!

h-vetinari commented:
> > I agree that this is getting rather tight for 1.11. Is there a good reason we couldn't take a bit more time and get it into 1.12 early?
> 
> Just that we have to pull the trigger at some point because I think this PR brings a lot of useful stuff to the table.

My point was not specific to the state of this PR, which I couldn't judge. I was just going through the PRs that are milestoned for 1.11 and since there was still ongoing discussion, I agreed with the caution about not merging in a hurry, on the general principle that it's better to do the review properly first, rather than have to fix something that's already published.

That said, if the people involved are comfortable with merging this, I have no objections of course!

tylerjereddy reviewed (commented):
I read through every comment in this PR, and studied the code for a while. Clearly, there is some consensus of quite thorough and high quality work here.

On the other hand, the diff is absolutely massive, and some core developers express concern about the "final" design of the API.

As release manager, I think my view here is this: I'll bump the milestone, but if one of the suitable FFT regulars wants to take responsibility for merging and maintaining the consequences, feel free to do so and reattach the milestone after merging.

larsoner commented:
Let's try to merge early in the 1.12 milestone then! 

@tupui do you want to look next to see if your comments have been sufficiently addressed? Looks like @tylerjereddy 's comments can be addressed simultaneously

vollbier commented:
Thank you so much for your helpful comments, @larsoner, @tupui, @tylerjereddy! I'm finished working through all of them and I resolved the conversations where there were obvious fixes. I left the ones open where my comments or fixed are not as straightforward.

endolith commented:
> @endolith also looked back in Nov 22 and was happy so that gives me some additional confidence.

I didn't look too in-depth, just skimmed through it and made some small suggestions.  I've been meaning to but haven't gotten around to it.

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+   definition to ``phase_shift = 0``), which suppresses linear phase components
+   for unshifted signals.
+   Furthermore, the FFT may be oversampled by padding :math:`w[m]` with zeros.
+   This can be achieved by specifying `mfft` to be larger that than the window
```
```suggestion
   This can be achieved by specifying `mfft` to be larger than the window
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+   canonical dual window of :math:`w[m]` and is also made up of :math:`M`
+   samples.
+
+Note that not for all windows and hops an inverse STFT exists. For a given
```
```suggestion
Note that an inverse STFT does not necessarily exist for all windows and hop sizes. For a given
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+incorporated inversion), so only components overlapping with :math:`w[m]` have
+an effect. Hence, all :math:`U_p[m, k]` far enough from the border are
+identical windows. To circumvent border effects, :math:`x[k]` is padded with
+zeros, enlarging :math:`\vb{U}` so all slices which touch :math:`x[k]` contain
+identical the dual window
```
```suggestion
zeros, enlarging :math:`\vb{U}` so all slices which touch :math:`x[k]` contain
the identical dual window
```

Is this what was meant?

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+
+showing that :math:`\vb{U}_p` and :math:`\vb{W}_{\!p}` are interchangeable.
+Hence, :math:`w_d[m]` is also a valid window with dual window :math:`w[m]`.
+Note that :math:`w_d[m]` is not a unique dual window, due :math:`\vb{s}`
```
```suggestion
Note that :math:`w_d[m]` is not a unique dual window, due to :math:`\vb{s}`
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+between the older "legacy" and the newer |ShortTimeFFT| implementations. The
+main motivation for a rewrite was the insight that integrating :ref:`dual
+windows <tutorial_stft_dual_win>` could not be done in a sane way without
+breaking compatability. This opened the opportunity for rethinking the code
```
```suggestion
breaking compatibility. This opened the opportunity for rethinking the code
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+    >>> t_lo, t_hi, f_lo, f_hi = SFT.extent(N, center_bins=True)
+    >>> t_str0 = r"Legacy stft() produces $%d\times%d$ points" % Sz0.T.shape
+    >>> t_str1 = r"ShortTimeFFT produces $%d\times%d$ points" % Sz1.T.shape
+    >>> _ = axx[0].set(title=t_str0, xlim=(t_lo, t_hi), ylim=(f_lo, f_hi))
+    >>> _ = axx[1].set(title=t_str1, xlabel="Time $t$ in seconds " +
```
```suggestion
    >>> axx[0].set(title=t_str0, xlim=(t_lo, t_hi), ylim=(f_lo, f_hi))
    >>> axx[1].set(title=t_str1, xlabel="Time $t$ in seconds " +
```

Is there a reason to assign to the throwaway `_`?  Suppressing doctest output or something?

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+section, all slices which touch the signal are incorporated in the new version.
+This has the advantage that the STFT can be sliced and reassembled as shown in
+the |ShortTimeFFT| code example. Furthermore, using all touching slices makes
+the ISTFT more robust in the case the windows are not non-zero everywhere.
```
```suggestion
the ISTFT more robust in the case of windows that are zero somewhere.
```

The double negative is a little confusing

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+    >>> np.allclose(Sz0, Sz1[:, 2:-1])
+    True
+
+Generally, those the additional slices contain non-zero values. Due to the
```
```suggestion
Generally, those additional slices contain non-zero values. Due to the
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+
+Note that the legacy implementation returns a signal, which is longer than the
+original. On the other hand, the new `istft` allows the length signal length or
+the slice to be specified. The deviation in length in the old implementation is
+caused by the fact the signal length is not a multiple of the slices.
```
```suggestion
the slice to be specified. The length discrepancy in the old implementation is
caused by the fact that the signal length is not a multiple of the slices.
```

"the new `istft` allows the signal length or the slice to be specified" meaning the k0, k1 parameter?

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+* The parameter ``fft_mode='centered'`` ensures that the zero frequency is
+  vertically centered for two-sided FFTs in the plot. With the legacy
+  implementation, `fftshift <scipy.fft.fftshift>` needs to be utilized.
+  ``fftyp='twosided'`` produces the same behavior as the old version.
```
```suggestion
  ``fft_mode='twosided'`` produces the same behavior as the old version.
```

Is this an old name that wasn't updated?

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+    >>> np.allclose(Sz2, Sz3)
+    True
+
+The difference to the other STFTs is that the time slices do not start at 0 but
```
```suggestion
The difference from the other STFTs is that the time slices do not start at 0 but
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+           4.625, 4.675, 4.725, 4.775, 4.825, 4.875])
+
+Furthermore, only slices which do not stick out to the right are returned,
+making the last slice be centered at 4.875 s, being shorter as with the default
```
```suggestion
centering the last slice at 4.875 s, which makes it shorter than with the default
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+For working with sampled signals :math:`x[k] := x(kT)`, :math:`k\in\IZ` with
+sampling interval :math:`T` (being the inverse of the sampling frequency `fs`),
+the discrete version, i.e., only evaluating the STFT at discrete grid points
+:math:`S[q, p] := S(q \Delta f, p\Delta t)`, :math:`q,p\in\IZ`, needs to be
```
Can you define what `p` and `q` are around here?

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+the ISTFT more robust in the case the windows are not non-zero everywhere.
+
+Note that the slices with identical time stamps produce equal results
+(up to numerical accuracy), i. e.:
```
```suggestion
(up to numerical accuracy), i.e.:
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+
+#. Extract the :math:`p`-th slice by windowing with the window :math:`w[m]`
+   made up of :math:`M` samples (see `m_num`) centered
+   at :math:`t[p] :=  p \Delta t = h T` (see `delta_t`).
+   I. e.,
```
```suggestion
   at :math:`t[p] :=  p \Delta t = h T` (see `delta_t`),
   i.e.,
```

endolith commented on scipy/signal/_short_time_fft.py:
```diff
@@ -0,0 +1,1666 @@
+    >>> f_i = 1 * np.arctan((t_x - t_x[N // 2]) / 2) + 5  # varying frequency
+    >>> x = np.sin(2*np.pi*np.cumsum(f_i)*T_x) # the signal
+
+    The utitlized Gaussian window is 50 samples or 2.5 s long. The parameter
```
```suggestion
    The utilized Gaussian window is 50 samples or 2.5 s long. The parameter
```

endolith commented on doc/source/tutorial/signal.rst:
```diff
@@ -1048,8 +1048,598 @@ giving a factor of :math:`\sim 2` speed increase over the straightforward
+Short-Time Fourier Transform
+----------------------------
+This section gives some background information on using the |ShortTimeFFT|
+class: For a complex valued signal :math:`x: \IR \mapsto\IC` the Short-time
```
Doesn't `\IR \mapsto\IC` mean it maps real signals to complex spectra?  But `ShortTimeFFT.stft(x...` says `The input signal as real or complex valued array`?

endolith replied:
This jumps immediately into rigorous math, could it just introduce the STFT in a basic way first to help people follow along?  Paraphrase "The STFT chops up the signal into overlapping chunks, tapers the ends of each, and applies FFT to each"

endolith commented:
I read through just the tutorial part last night and took some notes, still a lot more to read through.

vollbier commented:
Thanks @endolith, it always really helps, when someone else reads through my English...

vollbier commented:
I finally found the time to fully review all comments. Unless I overlooked something, I think I addressed all requested changes. As far as I can tell from the comments, all parts of this PR have been reviewed by someone. Thanks again @endolith, @larsoner, @tylerjereddy , @tupui for the great input and the patience for this lengthy PR. From my side, I see no obstacles for merging. 

PS: The failed test is unrelated to this PR (rebasing did not help).

tylerjereddy commented:
> Let's try to merge early in the 1.12 milestone then!

Ok, I may squash merge because there are a ton of commits, and if we find issues from this massive diff in the nightlies we'll want to be able to revert with a reasonably low amount of friction.

I'm not sure we'll ever get to the point where a bunch of core devs are fully happy that every corner of the 6,000 line diff has been studied, so I'll give this a shot based on a mix of positive reviews and questions about final API design that might get hammered out after merging.

One of the svg files is maybe a bit large? @vollbier I assume you're prepared to accept long-term maintenance duties on this contribution since you've been pushing for it :)

vollbier commented:
Of course, @tylerjereddy – I would not abandon a baby that was 9 month in the making. Thanks for merging.
Do you need a release notes entry?

PS: I'll take a look at the SVGs.

endolith commented:
Sorry, I keep meaning to read through the code and review it but haven't
had time.  I had some other comments from last time I looked at it, such as
lower_border_end and upper_border_begin seeming like a frequency bound when
they're actually more like a start and stop time bound?

On Sat, Jul 8, 2023 at 10:31 AM Dietrich Brunn ***@***.***>
wrote:

> I finally found the time to fully review all comments. Unless I overlooked
> something, I think I addressed all requested changes. As far as I can tell
> from the comments, all parts of this PR have been reviewed by someone.
> Thanks again @endolith <https://github.com/endolith>, @larsoner
> <https://github.com/larsoner>, @tylerjereddy
> <https://github.com/tylerjereddy> , @tupui <https://github.com/tupui> for
> the great input and the patience for this lengthy PR. From my side, I see
> no obstacles for merging.
>
> PS: The failed test is unrelated to this PR (rebasing did not help).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/17408#issuecomment-1627346869>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAOJ44PTIQPQHBPO7PE3WTXPFVMPANCNFSM6AAAAAAR6MOJXI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

vollbier commented:
Yes, it's the time axis.  I was not creative enough to come up with a more  expressive name for the shaded area in the plot, i.e., the time interval  where the window slices are not fully "inside" the signal. I named those intervals "lower border" and "upper border", which have both a beginning and a ending.
